pkg/sqlite/load.go

```diff
@@ -1536,3 +1547,46 @@ func (s *sqlLoader) deprecated(tx *sql.Tx, name string) (bool, error) {
 	// Ignore any deprecated bundles
 	return err == nil, err
 }
+
+// DeprecationAwareLoader understands how bundle deprecations are handled in SQLite and decorates
+// the sqlLoader with proxy methods that handle deprecation related table housekeeping.
+type DeprecationAwareLoader struct {
+	*sqlLoader
+}
+
+// NewDeprecationAwareLoader returns a new DeprecationAwareLoader.
+func NewDeprecationAwareLoader(db *sql.DB, opts ...DbOption) (*DeprecationAwareLoader, error) {
+	loader, err := newSQLLoader(db, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DeprecationAwareLoader{sqlLoader: loader}, nil
+}
+
+func (d *DeprecationAwareLoader) clearLastDeprecatedInPackage(pkg string) error {
+	tx, err := d.db.Begin()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		tx.Rollback()
+	}()
+
+	// The last deprecated bundles for a package will still have "tombstone" records in channel_entry (among other tables).
+	// Use that info to relate the package to a set of rows in the deprecated table.
+	_, err = tx.Exec(`DELETE FROM deprecated WHERE deprecated.operatorbundle_name IN (SELECT DISTINCT deprecated.operatorbundle_name FROM (deprecated INNER JOIN channel_entry ON deprecated.operatorbundle_name = channel_entry.operatorbundle_name) WHERE channel_entry.package_name = ?)`, pkg)
+	if err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+func (d *DeprecationAwareLoader) RemovePackage(pkg string) error {
```

Comment:
Created a decorator type because we don't want to drop the deprecation history in certain cases; e.g. `--overwrite-latest` needs to remember deprecations but remove everything else in the package.